### ISSUE TITLE
Complete IO hang on CentOS 7.5

### ISSUE THREAD

**Issue by fbezdeka** [nfs-ganesha/nfs-ganesha#335]
Hi!

the following problems occurs on each node of three node glusterfs cluster.
I'm exporting a glusterfs volume called "mirror" that is synced / mirrored between all nodes of the cluster.
The operating system is CentOS 7.5.
Nfs-ganesha version is 2.6.2

The problem occurs sometimes right after rebooting one of the machines, sometimes it appears after several hours of operation.

The logging message that comes up is "quite new" and was introduced with b5a9f5b6a5c04b96f02cf486ca792223998f13a2.

Please let me know if you need further details. 
Thanks!

Configuration:
```
EXPORT {
    Export_Id = 1 ;   # Export ID unique to each export
    Path = "/mirror_nfs";  # Path of the volume to be exported. Eg: "/test_volume"

    FSAL {
        name = GLUSTER;
        hostname = "127.0.0.1";  # IP of one of the nodes in the trusted pool
        volume = "mirror";  # Volume name. Eg: "test_volume"
    }

    CLIENT {
        Clients = 127.0.0.1, ::1;
        Access_Type = RW;
    }

    CLIENT {
        Clients = 0.0.0.0, ::0;
        Access_Type = NONE;
    }

    Squash = No_root_squash;
    Disable_ACL = TRUE; 
    Pseudo = "/mirror_nfs_pseudo"; 
    Protocols = "3","4" ; 
    Transports = "UDP","TCP" ; 
    SecType = "sys";
}
```

Log:
```
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1388[main] main :MAIN :EVENT :ganesha.nfsd Starting: Ganesha Version 2.6.2
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_set_param_from_conf :NFS STARTUP :EVENT :Configuration file successfully parsed
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] init_server_pkgs :NFS STARTUP :EVENT :Initializing ID Mapper.
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] init_server_pkgs :NFS STARTUP :EVENT :ID Mapper successfully initialized.
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] glusterfs_create_export :FSAL :EVENT :Volume mirror exported at : '/'
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] lower_my_caps :NFS STARTUP :EVENT :CAP_SYS_RESOURCE was successfully removed for proper quota management in FSAL
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] lower_my_caps :NFS STARTUP :EVENT :currenty set capabilities are: = cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap+ep
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_start_grace :STATE :EVENT :NFS Server Now IN GRACE, duration 90
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_Init_svc :DISP :CRIT :Cannot acquire credentials for principal nfs
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_Init_admin_thread :NFS CB :EVENT :Admin thread initialized
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_rpc_cb_init_ccache :NFS STARTUP :EVENT :Callback creds directory (/var/run/ganesha) already exists
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_rpc_cb_init_ccache :NFS STARTUP :WARN :gssd_refresh_krb5_machine_credential failed (-1765328160:22)
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_Start_threads :THREAD :EVENT :Starting delayed executor.
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_Start_threads :THREAD :EVENT :9P/TCP dispatcher thread was started successfully
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_Start_threads :THREAD :EVENT :gsh_dbusthread was started successfully
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_Start_threads :THREAD :EVENT :admin thread was started successfully
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[_9p_disp] _9p_dispatcher_thread :9P DISP :EVENT :9P dispatcher started
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_Start_threads :THREAD :EVENT :reaper thread was started successfully
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_Start_threads :THREAD :EVENT :General fridge was started successfully
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_start :NFS STARTUP :EVENT :-------------------------------------------------
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_start :NFS STARTUP :EVENT :             NFS SERVER INITIALIZED
26/08/2018 12:55:48 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[main] nfs_start :NFS STARTUP :EVENT :-------------------------------------------------
26/08/2018 12:56:51 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[svc_15] mdcache_lru_fds_available :INODE LRU :CRIT :FD Hard Limit Exceeded, waking LRU thread.
26/08/2018 12:56:51 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[svc_12] mdcache_lru_fds_available :INODE LRU :CRIT :FD Hard Limit Exceeded, waking LRU thread.
26/08/2018 12:57:06 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[cache_lru] lru_run :INODE LRU :CRIT :Futility count exceeded.  The LRU thread is unable to make progress in reclaiming FDs, will try harder.
26/08/2018 12:57:06 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[cache_lru] lru_run :INODE LRU :CRIT :Futility count exceeded.  The LRU thread is unable to make progress in reclaiming FDs, will try harder.
26/08/2018 12:57:07 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[cache_lru] lru_run :INODE LRU :CRIT :Futility count exceeded.  The LRU thread is unable to make progress in reclaiming FDs, will try harder.
26/08/2018 12:57:11 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[cache_lru] lru_run :INODE LRU :CRIT :Futility count exceeded.  The LRU thread is unable to make progress in reclaiming FDs, will try harder.
26/08/2018 12:57:12 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[cache_lru] lru_run :INODE LRU :CRIT :Futility count exceeded.  The LRU thread is unable to make progress in reclaiming FDs, will try harder.
26/08/2018 12:57:16 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[cache_lru] lru_run :INODE LRU :CRIT :Futility count exceeded.  The LRU thread is unable to make progress in reclaiming FDs, will try harder.
26/08/2018 12:57:17 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[cache_lru] lru_run :INODE LRU :CRIT :Futility count exceeded.  The LRU thread is unable to make progress in reclaiming FDs, will try harder.
26/08/2018 12:57:18 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[reaper] nfs_lift_grace_locked :STATE :EVENT :NFS Server Now NOT IN GRACE
26/08/2018 12:57:20 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[svc_4] glusterfs_setattr2 :FSAL :CRIT :setattrs failed with error Stale file handle
26/08/2018 12:57:20 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[svc_4] fsal_setattr :FSAL :EVENT :FSAL returned STALE from setattr2
26/08/2018 12:57:21 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[cache_lru] lru_run :INODE LRU :CRIT :Futility count exceeded.  The LRU thread is unable to make progress in reclaiming FDs, will try harder.
26/08/2018 12:57:22 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[cache_lru] lru_run :INODE LRU :CRIT :Futility count exceeded.  The LRU thread is unable to make progress in reclaiming FDs, will try harder.
26/08/2018 12:57:23 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[cache_lru] lru_run :INODE LRU :CRIT :Futility count exceeded.  The LRU thread is unable to make progress in reclaiming FDs, will try harder.
26/08/2018 12:57:26 : epoch 5b828734 : _servername_ : ganesha.nfsd-1481[cache_lru] lru_run :INODE LRU :CRIT :Futility count exceeded.  The LRU thread is unable to make progress in reclaiming FDs, will try harder.
... last line is repeated endless ...
```

**Comment by dang** [comment]
What is the workload that's running on the client(s)?

**Comment by fbezdeka** [comment]
All affected systems are web servers serving PHP applications like Wordpress, Owncloud and Nextcloud.

There is no significant load on the systems.
A loadbalancer (triggering all servers every 2 seconds for fetching system status) is sufficient for reproduction.

I'm mounting the gluster volume using the NFS way due to performance reasons.
Especially Nextcloud is very slow when running on a native gluster volume mount.

**Comment by dang** [comment]
This message has been reduced to a warning in 2.7, since it's not as meaningful as it used to be.  (commit 7e1ed819d271949bcaae2b56694e45be610ca553)  This particular commit was not backported to 2.6.3.

However, it seems odd that this message would trigger with just a load balancer touching every 2 seconds.  What it means is that files are opened faster than the background thread can close them.   But with that kind of minimal load, you shouldn't even be over the high water mark.  What's the file limit for Ganesha?  (ulimit -n)

Note that the standard init scripts shipped with Ganesha set the file limit to 1048576, so hitting the high water mark under no load would be very strange.

**Comment by fbezdeka** [comment]
/lib/systemd/system/nfs-ganesha.service
```
...
[Service]
Type=forking
Environment="NOFILE=1048576"
EnvironmentFile=-/run/sysconfig/ganesha
ExecStart=/bin/bash -c "${NUMACTL} ${NUMAOPTS} /usr/bin/ganesha.nfsd ${OPTIONS} ${EPOCH}"
ExecStartPost=-/bin/bash -c "prlimit --pid $MAINPID --nofile=$NOFILE:$NOFILE"
ExecStartPost=-/bin/bash -c "/usr/bin/sleep 2 && /bin/dbus-send --system   --dest=org.ganesha.nfsd --type=method_call /org/ganesha/nfsd/admin  org.ganesha.nfsd.admin.init_fds_limit"
ExecReload=/bin/kill -HUP $MAINPID
ExecStop=/bin/dbus-send --system   --dest=org.ganesha.nfsd --type=method_call /org/ganesha/nfsd/admin org.ganesha.nfsd.admin.shutdown
...
```

cat /proc/%pid_of_ganesha.nfsd%/limits 
```
Limit                     Soft Limit           Hard Limit           Units     
Max cpu time              unlimited            unlimited            seconds   
Max file size             unlimited            unlimited            bytes     
Max data size             unlimited            unlimited            bytes     
Max stack size            8388608              unlimited            bytes     
Max core file size        0                    unlimited            bytes     
Max resident set          unlimited            unlimited            bytes     
Max processes             23129                23129                processes 
Max open files            1048576              1048576              files     
Max locked memory         65536                65536                bytes     
Max address space         unlimited            unlimited            bytes     
Max file locks            unlimited            unlimited            locks     
Max pending signals       23129                23129                signals   
Max msgqueue size         819200               819200               bytes     
Max nice priority         0                    0                    
Max realtime priority     0                    0                    
Max realtime timeout      unlimited            unlimited            us
```

So limits set by systemd unit file are beeing used.

Could the thread releasing this file descriptors crash somehow?
I will try to find out if php-fpm (or something else) will open so much files at once.

**Comment by dang** [comment]
If any thread crashes, all of Ganesha crashes.  That would mean >900k anonymous FDs open, which seems like a lot.  It's possible there's a FD leak somewhere, but without more information about the workload, it'll be hard to track down.

**Comment by dang** [comment]
Wait, I think I missed something.  Does I/O actually stop?  Or is the problem just the error messages in the log?  Because those error messages are unrelated to I/O on existing, open files.  It will keep new files from being opened.

**Comment by fbezdeka** [comment]
Yes, there is a complete IO hang on the NFS mount point.
All processes trying to read or write something from/to the NFS are in "ioblocked" state.

As a workarround I removed the NFS mount and replaced it by a native glusterfs mount.
The problem has gone now, but performance decreased a lot.

As far as I can see, there are a couple of processes waiting for IO.
Apache waits for php-fpm and php-fpm waits for the filesystem to deliver something.

I will try to fetch the number of files opened by ganesha.nfsd when the situation appears again.
You mentioned missing information about the workload.
What exactly do you need? I will try to provide whatever is needed...

**Comment by dang** [comment]
Ideally enough that we can reproduce it.  It appears something is sucking up FDs, but what causes that I don't yet know.   If we can characterize the load well enough, I can reproduce here, and debug the cause.

Unfortunately, checking the number of FDs open in the system won't help, since FSAL_GLUSTER doesn't use system FDs; instead, is uses GFAPI FDs.  So the open count in the OS won't go up.

**Comment by beeksfxpj** [comment]
Was there a resolution to this issue?   We are having the exact same issue.   New deployment of 2 gluster nodes with nfs-ganesha exporting a NFS mount point to VMWare ESX server.   With just one VM running and basically sitting idle the NFS mount point stops responding after a period of time and the nfs-ganesha service needs to be restarted to recover. 

nfs-ganesha-2.6.2-1
glusterfs-server-4.1.3
CentOS Linux release 7.5.1804


Our plan is downgrade gluster and use its native nfs ability if there isn't a fix.   

Would testing nfs-ganesha 2.7 be worth while?  Any fixes in it and is it stable yet?

**Comment by dang** [comment]
2.6.3 has a lot of bug fixes; I don't remember if any are related to open FD handling.  2.7 is quite stable and about to be released (in the next week).  So, either of those would be a good thing to test.

**Comment by fbezdeka** [comment]
Good to know that this problem occurs on other setups as well.

I already noticed the available glusterfs update and already installed it.
But I did not revert the workaround yet.
This will hopefully happen within the next days. Will keep you informed.

**Comment by beeksfxpj** [comment]
We are testing with nfs-ganesha 2.6.3 now.   So far so good, but need to give it more time.

**Comment by beeksfxpj** [comment]
The upgrade to 2.6.3 appears to have fixed the issue for us.  No problems since the upgrade and pretty impressive performance.

**Comment by beeksfxpj** [comment]
The issue came back for us just took a lot longer.   We have one node experiencing this currently.   What information can I post to help debug?

**Comment by pasikarkkainen** [comment]
@beeksfxpj : does the io hang happen randomly, or do you have a way to reproduce it ? what's the workload triggering it?

**Comment by beeksfxpj** [comment]
Just randomly.  There was absolutely no workload on it.  Sitting idle and we started getting the hang on all I/O.    I did as much stress testing as I could for a couple days, then left it alone for a few days when it.

**Comment by pasikarkkainen** [comment]
@dang any suggestions how to further debug/troubleshoot this issue? Thanks.

**Comment by fbezdeka** [comment]
Gluster 4.1.5 was released. The changelog mentions this bug: 
https://bugzilla.redhat.com/show_bug.cgi?id=1625575

Maybe @dang can tell us if this may be part of the problem we see on our setups.

References:
Changelog: https://docs.gluster.org/en/latest/release-notes/4.1.5/

**Comment by pasikarkkainen** [comment]
@fbezdeka I'd say it's good to try upgrading to gluster 4.1.5, and check if that changes anything. Doesn't hurt (much) :)

**Comment by dang** [comment]
I don't know, I'm sorry.  I'm far from an expert on Gluster.  It seems somewhat unlikely that replica count would change, or that a brick was being replaced.

**Comment by fbezdeka** [comment]
Updated the cluster to gluster 4.1.5, but the same problem comes up. Some tests are still missing, but I think there is no workload necessary to reproduce the the problem.

**Comment by pasikarkkainen** [comment]
@fbezdeka Ok. Can you increase loglevel? Maybe we can get more verbose logs to figure out what goes wrong..

**Comment by thotz** [comment]
You can also take the staedump of glusterfs client(nfs-ganesha) and and bricks as well
gluster v statedump <volume name> client localhost:$(pgrep ganesha) on the ganesha server
 
Also please check the pstack of ganesha process
@fbezdeka

**Comment by fbezdeka** [comment]
Summary of last week:

State when starting to investigate:
When switching to the NFS mount by editing `/etc/fstab` and then rebooting the system was completely unusable. Some warnings about "httpd was blocked for more than 120 seconds" could be found in the system log. `htop` and similar tools reported a very high load but actually only a few processes were producing cpu load. Most processes were in "disk blocked" state.

The first thing I was wondering about was the httpd state. 
I expected php-fpm to produce the IO load and not httpd. After a high number of tries to `strace` the httpd process (I had to reduce the number of worker processes and threads first) I was able to see some "hang" while httpd tried to write something into its logfile.

I rechecked httpd logging configuration on our system and was surprised to find out that all httpd instances of the affected cluster were actually writing into the same file for each vhost. The biggest logfile at this point was about 1,3 GB in size. Just to give it a chance I deleted this file and rebootet the system.

The system came up and no further problems could be discovered during the first 3 hours.

Within the first 3 hours I adjusted httpd and php-fpm configurations on all cluster nodes to use different logfiles for each node/vhost.

The problem with IO hangs came back on one node after some hours. We restarted nfs-ganesha service and the problem was gone. The cluster is in a good shape now for a few days. I will keep you informed.

One strange thing: The size of the log files seems to be significant. httpd is able to write into the same file for some time/size but when this limit is reached the hang comes up.

**Comment by pasikarkkainen** [comment]
@fbezdeka @beeksfxpj Can you please do the statedumps and the pstack when you experience the hang the next time.. as requested by @thotz . Thanks!

**Comment by fbezdeka** [comment]
The problem came back (on two of three nodes) after some days.

pstack: [pstack.txt](https://github.com/nfs-ganesha/nfs-ganesha/files/2453805/pstack.txt)

glusterdump: [glusterdump.1476.dump.1538907054.txt](https://github.com/nfs-ganesha/nfs-ganesha/files/2453806/glusterdump.1476.dump.1538907054.txt)

**Comment by thotz** [comment]
From the statedump and pstack it looks like nothing stuck at glusterfs stack. is the issue related nfs client side?
@fbezdeka can u please check anything got stuck nfs-client stack ?

**Comment by fbezdeka** [comment]
> @fbezdeka can u please check anything got stuck nfs-client stack ?

@thotz: Can you tell me how to do that?

**Comment by thotz** [comment]
Even I don't much expertise with nfs-client
There will proc directory for the nfs-client-module, go inside of that directory there are files like status, stack and directories like fd/fd_info etc. You can check those

U can also turn on nfs -client level debugging using following

rpcdebug -m nfs -s all  --enable

rpcdebug -m nfs -c all -- disable

**Comment by pasikarkkainen** [comment]
Weird that both the linux nfs client and ESXi nfs client seem to "hang", but nothing obvious on the nfs-ganesha server dumps..

Let's hope the status info from the nfs client helps.

**Comment by fbezdeka** [comment]
After some hours of operation the following appears in the syslog:

```
Okt 22 21:11:12 hostname kernel: NFS: permission(0:41/13063406188894154770), mask=0x10, res=0
Okt 22 21:11:12 hostname kernel: NFS: dentry_delete(vhost/footer.php, 40808cc)
Okt 22 21:11:12 hostname kernel: NFS: nfs_weak_revalidate: inode 10332493177315662170 is valid
Okt 22 21:11:12 hostname kernel: NFS: permission(0:41/10332493177315662170), mask=0x41, res=0
Okt 22 21:11:13 hostname kernel: NFS: nfs_pgio_result: 41836, (status -528)
Okt 22 21:11:14 hostname kernel: NFS: nfs_pgio_result: 43924, (status -528)
Okt 22 21:11:14 hostname kernel: NFS: nfs_pgio_result: 43923, (status -528)
Okt 22 21:11:16 hostname kernel: NFS: nfs_pgio_result: 41797, (status -528)
Okt 22 21:11:18 hostname kernel: NFS: nfs_pgio_result: 41836, (status -528)
Okt 22 21:11:19 hostname kernel: NFS: nfs_pgio_result: 43924, (status -528)
``` 
Restarting nfs-ganesha still helps...

**Comment by dang** [comment]
Those messages are from knfsd, and not from Ganesha at all.  Is something starting knfsd behind your back?

**Comment by fbezdeka** [comment]
I enabled the debugging flags as requested using

```
# rpcdebug -m nfs -s all 
```

The kernel nfs server is not running.

```
# ps aux | grep nfs
root      1423  0.0  0.0      0     0 ?        S<   Sep30   0:00 [nfsiod]
root      2778  0.6  1.2 3612676 73536 ?       Ssl  09:24   0:04 /usr/bin/ganesha.nfsd -L /var/log/ganesha/ganesha.log -f /etc/ganesha/ganesha.conf -N NIV_EVENT
```

```
# service nfs status
Redirecting to /bin/systemctl status nfs.service
● nfs-server.service - NFS server and services
   Loaded: loaded (/usr/lib/systemd/system/nfs-server.service; disabled; vendor preset: disabled)
   Active: inactive (dead)
```

**Comment by fbezdeka** [comment]
Some more log entries coming up when the hang occurs:

```
Okt 28 09:53:23 hostname kernel: nfs: server localhost not responding, still trying
Okt 28 09:53:23 hostname kernel: nfs: server localhost not responding, still trying
Okt 28 09:53:35 hostname kernel: nfs: server localhost not responding, still trying
...
```

Seems the nfs client losses the connection to ganesha.

**Comment by fbezdeka** [comment]
I went one step further. 
For some reason the glusterfs logs on our systems are using wrong timezone data.
Took some time to figure that out...

1. Ganesha looses the connection to glusterfs for some reason:
```
Nov 11 10:34:44 host3 ganesha.nfsd[27405]: [2018-11-11 09:34:44.222805] C [rpc-clnt-ping.c:166:rpc_clnt_ping_timer_expired] 0-mirror-client-2: server %ip_of_host3%:49152 has not responded in the last 42 seconds, disconnecting.
```

2. Kernel (NFS client) losses the connection to ganesha:
```
...
Nov 11 10:36:51 host3 kernel: nfs: server localhost not responding, still trying
Nov 11 10:36:51 host3 kernel: nfs: server localhost not responding, still trying
...
```

3. GlusterFS logs:
```
[2018-11-11 09:34:44.223550] I [MSGID: 115036] [server.c:483:server_rpc_notify] 0-mirror-server: disconnecting connection from CTX_ID:68243179-52e3-4a11-a7ff-b9645f0d3ade-GRAPH_ID:0-PID:27405-HOST:host3-PC_NAME:mirror-client-2-RECON_NO:-0
[2018-11-11 09:34:44.223615] I [MSGID: 115013] [server-helpers.c:286:do_fd_cleanup] 0-mirror-server: fd cleanup on /home/vhost/logs/host3.www.vhost.com.access.log
[2018-11-11 09:34:44.223710] I [MSGID: 115013] [server-helpers.c:286:do_fd_cleanup] 0-mirror-server: fd cleanup on /home/vhost/php_sessions/sess_6t1ca2pi39jn53qj56j7q3jgkf
[2018-11-11 09:34:44.223729] I [MSGID: 115013] [server-helpers.c:286:do_fd_cleanup] 0-mirror-server: fd cleanup on /home/vhost/php_sessions/sess_jsj9ac35p6koe0snb6mk6g15m5
[2018-11-11 09:34:44.223747] I [MSGID: 115013] [server-helpers.c:286:do_fd_cleanup] 0-mirror-server: fd cleanup on /home/vhost/php_sessions/sess_u5j89mbmrqbhisdmuc1cjnevi1
[2018-11-11 09:34:44.223764] I [MSGID: 115013] [server-helpers.c:286:do_fd_cleanup] 0-mirror-server: fd cleanup on /home/vhost/php_sessions/sess_dud95t1qj5hlio49c08h6et3n2
[2018-11-11 09:34:44.223783] I [MSGID: 115013] [server-helpers.c:286:do_fd_cleanup] 0-mirror-server: fd cleanup on /home/vhost/php_sessions/sess_eg1q2etj0sk6hk837r1fp8qj8r
[2018-11-11 09:34:44.223799] I [MSGID: 115013] [server-helpers.c:286:do_fd_cleanup] 0-mirror-server: fd cleanup on /home/vhost/php_sessions/sess_1mdbnpem20isjkof0p0f140oju
[2018-11-11 09:34:44.224308] I [MSGID: 101055] [client_t.c:444:gf_client_unref] 0-mirror-server: Shutting down connection CTX_ID:68243179-52e3-4a11-a7ff-b9645f0d3ade-GRAPH_ID:0-PID:27405-HOST:host3-PC_NAME:mirror-client-2-RECON_NO:-0
```

When I look at the glusterfs logfiles while (re)starting ganesha, I can see that ganesha establishes a tcp connection to each affected glusterfs server (and not only to "127.0.0.1" as configured in the config, see first comment). For some reason (that I don't understand) the connection to the glusterfs daemon running on the same host times out. So both (ganesha + glusterfs) are closing this connection.

Is there some configuration option available that would tell ganesha just to reconnect on failure?

The native glusterfs mount keeps reachable/accessable when ganesha losses the connection.

**Comment by fbezdeka** [comment]
Some more findings:

When the system is in "hang" state, I'm not able to do any socket communication to local services.
```
telnet <localip, e.g. 192.168.x.x on eth0> 80
```
will hang until I restart nfs-ganesha.

Seems that this lines are coming from the kernel right before the "crash" occurs:
```
[2114593.244423] nfs: server localhost OK
[2114593.253421] nfs: server localhost OK
[2119458.723302] NFS: directory lib/l10n contains a readdir loop.Please contact your server vendor.  The file: pt_BR.json has duplicate cookie 299
[2119458.723358] NFS: directory lib/l10n contains a readdir loop.Please contact your server vendor.  The file: pt_BR.json has duplicate cookie 299
```

**Comment by fbezdeka** [comment]
We updated the kernel of one affected system to the latest LTS version (4.19.1) end enabled some debug options. The idea was to make sure it is not the NFS client that causes this issue.

I think it will not help much, but to be complete: 
The following (kernel) call trace came up:
```
[ 5162.334666] INFO: task php-fpm:20974 blocked for more than 120 seconds.
[ 5162.336101]       Not tainted 4.19.1-1.el7.elrepo.x86_64 #1
[ 5162.337111] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[ 5162.338280] php-fpm         D    0 20974   3528 0x00000080
[ 5162.338284] Call Trace:
[ 5162.338311]  __schedule+0x2ab/0x880
[ 5162.338319]  ? __d_alloc+0x130/0x240
[ 5162.338321]  schedule+0x36/0x80
[ 5162.338322]  d_alloc_parallel+0x385/0x500
[ 5162.338332]  ? wake_up_q+0x80/0x80
[ 5162.338334]  __lookup_slow+0x6e/0x150
[ 5162.338336]  lookup_slow+0x3a/0x60
[ 5162.338337]  walk_component+0x163/0x250
[ 5162.338344]  ? security_inode_permission+0x3f/0x60
[ 5162.338346]  link_path_walk+0x1b3/0x620
[ 5162.338347]  path_openat+0xe9/0x1400
[ 5162.338359]  ? do_async_page_fault+0x22/0xd0
[ 5162.338369]  do_filp_open+0x91/0x100
[ 5162.338376]  ? strncpy_from_user+0x5c/0x160
[ 5162.338381]  ? __alloc_fd+0xae/0x170
[ 5162.338386]  do_sys_open+0x124/0x210
[ 5162.338388]  __x64_sys_open+0x21/0x30
[ 5162.338393]  do_syscall_64+0x60/0x190
[ 5162.338396]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[ 5162.338398] RIP: 0033:0x7f8a9e0bfa30
[ 5162.338404] Code: Bad RIP value.
[ 5162.338405] RSP: 002b:00007fff714a7758 EFLAGS: 00000246 ORIG_RAX: 0000000000000002
[ 5162.338410] RAX: ffffffffffffffda RBX: 0000000000000003 RCX: 00007f8a9e0bfa30
[ 5162.338411] RDX: 000000000000000a RSI: 0000000000000000 RDI: 00007f8a9e1588ee
[ 5162.338411] RBP: 00007fff714a7770 R08: 00007f8a9ddcc300 R09: 00000000000000ca
[ 5162.338412] R10: 00007f8a9ddcc300 R11: 0000000000000246 R12: 00000000000007d1
[ 5162.338413] R13: 0000000000000000 R14: 000000000000000a R15: 0000558629e98b43
```

As always: Restarting the ganesha-nfs service helps.

**Comment by dang** [comment]
Okay, this appears to be a readdir bug.  Can you try with 2.7.1?  It has some readdir fixes that have not been backported to 2.6.

**Comment by fbezdeka** [comment]
We updated to 2.7.1 but same problem...

**Comment by dang** [comment]
Okay, this is something else, not a readdir error.  So, progress, of a sort.  The issue now is that something is wrong with FD accounting.  I suspect this is related to the crash (a lock was taken that was was already held by that thread)  Do you have a backtrace of the crashed system?  That would help a lot.

**Comment by fbezdeka** [comment]
I attached gdb to the already running nfs-ganesha process.

```
The program being debugged has been started already.
Start it from the beginning? (y or n) n
Program not restarted.
(gdb)
(gdb)
(gdb) c
Continuing.
[Thread 0x7f7bd428d700 (LWP 15468) exited]
[New Thread 0x7f7bd61d7700 (LWP 15513)]
[Thread 0x7f7bd61d7700 (LWP 15513) exited]
[New Thread 0x7f7b186f8700 (LWP 15535)]
[New Thread 0x7f7bd61d7700 (LWP 15613)]
[New Thread 0x7f7bd428d700 (LWP 15614)]

Program received signal SIGABRT, Aborted.
[Switching to Thread 0x7f7bd5ed4700 (LWP 14962)]
0x00007f7bee876277 in raise () from /lib64/libc.so.6
(gdb) bt
#0  0x00007f7bee876277 in raise () from /lib64/libc.so.6
#1  0x00007f7bee877968 in abort () from /lib64/libc.so.6
#2  0x000055debf4f34c8 in mdc_clean_entry (entry=entry@entry=0x55dec3447290) at /usr/src/debug/nfs-ganesha-2.7.1/src/FSAL/Stackable_FSALs/FSAL_MDCACHE/mdcache_helpers.c:231
#3  0x000055debf4e337a in mdcache_lru_clean (entry=0x55dec3447290) at /usr/src/debug/nfs-ganesha-2.7.1/src/FSAL/Stackable_FSALs/FSAL_MDCACHE/mdcache_lru.c:591
#4  _mdcache_lru_unref (entry=entry@entry=0x55dec3447290, flags=flags@entry=0, func=func@entry=0x55debf53f6f0 <__func__.24087> "cih_remove_checked", line=line@entry=406)
    at /usr/src/debug/nfs-ganesha-2.7.1/src/FSAL/Stackable_FSALs/FSAL_MDCACHE/mdcache_lru.c:1989
#5  0x000055debf4fa945 in cih_remove_checked (entry=0x55dec3447290) at /usr/src/debug/nfs-ganesha-2.7.1/src/FSAL/Stackable_FSALs/FSAL_MDCACHE/mdcache_hash.h:406
#6  _mdcache_kill_entry (entry=entry@entry=0x55dec3447290, file=file@entry=0x55debf53d150 "/builddir/build/BUILD/nfs-ganesha-2.7.1/src/FSAL/Stackable_FSALs/FSAL_MDCACHE/mdcache_file.c", line=line@entry=549,
    function=function@entry=0x55debf53d31d <__func__.21108> "mdc_read_cb") at /usr/src/debug/nfs-ganesha-2.7.1/src/FSAL/Stackable_FSALs/FSAL_MDCACHE/mdcache_helpers.c:3269
#7  0x000055debf4f043d in mdc_read_cb (obj=<optimized out>, ret=..., obj_data=<optimized out>, caller_data=0x7f7bac009bc0) at /usr/src/debug/nfs-ganesha-2.7.1/src/FSAL/Stackable_FSALs/FSAL_MDCACHE/mdcache_file.c:549
#8  0x00007f7bea6a4c69 in glusterfs_read2 (obj_hdl=0x55dec34471b8, bypass=<optimized out>, done_cb=0x55debf4f0360 <mdc_read_cb>, read_arg=0x7f7bd5ed2b90, caller_arg=0x7f7bac009bc0)
    at /usr/src/debug/nfs-ganesha-2.7.1/src/FSAL/FSAL_GLUSTER/handle.c:1947
#9  0x000055debf4f1784 in mdcache_read2 (obj_hdl=0x55dec34472c8, bypass=<optimized out>, done_cb=<optimized out>, read_arg=0x7f7bd5ed2b90, caller_arg=<optimized out>)
    at /usr/src/debug/nfs-ganesha-2.7.1/src/FSAL/Stackable_FSALs/FSAL_MDCACHE/mdcache_file.c:584
#10 0x000055debf454774 in nfs3_read (arg=<optimized out>, req=<optimized out>, res=0x7f7bac0090e0) at /usr/src/debug/nfs-ganesha-2.7.1/src/Protocols/NFS/nfs3_read.c:311
#11 0x000055debf417e9f in nfs_rpc_process_request (reqdata=0x7f7bac002000) at /usr/src/debug/nfs-ganesha-2.7.1/src/MainNFSD/nfs_worker_thread.c:1329
#12 0x000055debf41734a in nfs_rpc_decode_request (xprt=0x7f7bc0000c10, xdrs=0x7f7bac00fb70) at /usr/src/debug/nfs-ganesha-2.7.1/src/MainNFSD/nfs_rpc_dispatcher_thread.c:1341
#13 0x00007f7bf0a8366d in svc_rqst_xprt_task () from /lib64/libntirpc.so.1.7
#14 0x00007f7bf0a839ba in svc_rqst_run_task () from /lib64/libntirpc.so.1.7
#15 0x00007f7bf0a8b87b in work_pool_thread () from /lib64/libntirpc.so.1.7
#16 0x00007f7bef032e25 in start_thread () from /lib64/libpthread.so.0
#17 0x00007f7bee93ebad in clone () from /lib64/libc.so.6
```

**Comment by dang** [comment]
So, this is the attr_lock it's crashing on, while tearing down the entry.  Nothing in that call path will take the lock, so it doesn't seem that it can be a double-lock.  This means it's possibly a use-after-free, but this is the normal destroy-on-unref path, which is exercised all the time, so it should be solid.  I've looked over this whole thing, and I can't find what might cause this.  @ffilz any ideas?

**Comment by ffilz** [comment]
> So, this is the attr_lock it's crashing on, while tearing down the entry. Nothing in that call path will take
> the lock, so it doesn't seem that it can be a double-lock. This means it's possibly a use-after-free, but
> this is the normal destroy-on-unref path, which is exercised all the time, so it should be solid. I've
> looked over this whole thing, and I can't find what might cause this. @ffilz any ideas?

Any possibility a previous request leaked an attr_lock so this thread just happens to have the attr_lock held?

Also, why does ERR_FSAL_DELAY result in killing the object anyway?

**Comment by dang** [comment]
Unlikely.  It would have to have leaked this same entry's attr_lock, and would have to have been scheduled on the same thread.  And, if we had an attr_lock leak, it would have shown up long before this as a deadlock, at least.

As for DELAY, it's been that way all along.

**Comment by dang** [comment]
The configured gluster machine only tells Ganesha how to connect to the cluster.  Once it's done so, it downloads the cluster config, and then  GFAPI (the library Ganesha uses to connect to Gluster) will connect to the entire cluster as necessary, based on the actual layout of the volume in question.  I'm not an expert on Gluster, but my understanding is that for a mirrored volume, a client will connect to all the bricks and spread out the I/O across them.

This doesn't explain why it can't use one of the other bricks when one goes down; I don't know enough about Gluster internals to answer that.

**Comment by fbezdeka** [comment]
No. I only used a setup with three nodes.

**Comment by pasikarkkainen** [comment]
Ok. Just trying to think of something to narrow down this issue.

**Comment by tuhoanganh** [comment]
I think i have the same problems with @fbezdeka 
Version:
- nfs-ganesha-2.6.3-1.el7.x86_64
- glusterfs-4.1.5-1.el7.x86_64

I'm using 2 node replication gluser and nfs-ganesha for mouting glusterfs volume running on the same host. eg:
- node 1 mount gluster volume on node 1
- node 2 mount gluster volume on node 2

When i try to shutdown one of nodes and boot it up, i cannot view any file in nfs folder with cat command...

I alse tried restart nfs-ganesha service and wait... sometime it works (several minutes after restating process done) and sometime not.

**Comment by soumyakoduri** [comment]
I dont know the entire context. But recently I found a potential deadlock which could happen with upcall notifications sent while nfs-ganesha server is coming up leading to server unresponsiveness. Its fixed with https://review.gluster.org/21656 .

Could you turn off upcalls (if enabled) and check
"gluster v set  <volname> features.cache-invalidation off"

**Comment by fbezdeka** [comment]
features.cache-invalidation was on.
Switched to off now. Let's see...

**Comment by indi81** [comment]
I've had exactly the same scenario happening here. Running Gluster, using nfs-ganesha with localhost mount. NGINX and PHP-FPM. Wordpress application placed in NFS-ganesha-mounted path.

Also set the _features.cache-invalidation off_ as a test now, it was turned on here as well.

CentOS 7.5
GlusterFS 3.12.10
NFS-Ganesha 2.5.5

Locks always resolved by restarting ganesha.

**Comment by fbezdeka** [comment]
I have good news so far!
Disabling _features.cache-invalidation_ fixed my locking/hang problem.
The system has been under test for nearly two weeks now. 
The problem did not come back yet.

It seems that we lost some performance, but that is "acceptable" for now.

**Comment by fbezdeka** [comment]
@soumyakoduri 
Which versions of gluster will contain your fix?

**Comment by soumyakoduri** [comment]
The fix is available in glusterfs mainline branch and in glusterfs-5.2 release . Its yet to be backported to glusterfs-4.x release.

Also please note that this particular cache-invalidation feature (option "features.cache-invalidation") is needed only if you have multiple nfs-ganesha servers serving same glusterfs volume. Otherwise you can keep the option off. in your workloads.

**Comment by indi81** [comment]
Our problem came back today I am afraid. I am unsure if we restarted the nodes after disabling features.cache-invalidation so we did that as a follow up now.

We are using a two node cluster.

**Comment by pasikarkkainen** [comment]
@fbezdeka : How's your ganesha now, still working OK after disabling features.cache-invalidation ?

And Happy 2019 to everyone! :)

**Comment by soumyakoduri** [comment]
The fix is merged and available in all gluster current stable releases. There is no need to change network.ping-timeout value. 

by " the other node still hangs for a moment."  , do you mean nfs-ganesha is not responding for a while and then becomes responsive? That is because when any node goes down and failover happens, all other ganesha nodes are put in grace. This will block I/Os till the server comes out of grace period which is 90sec by default.

**Comment by soumyakoduri** [comment]
yes..you can change it to the desired value using below ganesha option -

NFSV4 {
Grace_Period = 90;
}

**Comment by tuhoanganh** [comment]
> yes..you can change it to the desired value using below ganesha option -
> 
> NFSV4 {
> Grace_Period = 90;
> }

Thank you very much,
It works like a charm :D.

**Comment by fbezdeka** [comment]
I updated to the now available gluster packages and reenabled _features.cache-invalidation_.
The problem came back :sob: 
The situation is even worse: I'm unable to restart ganesha (timeout) on a hanging system.

Going to disable _features.cache-invalidation_ as workarround again.

**Comment by soumyakoduri** [comment]
Sorry. I hadn't noticed your latest comment. Could you please collect and provide the stack trace  of ganesha process when it hangs using below command - 
#gstack <nfs-ganesha-pid> 

Collect stack traces for about 2-3 times giving few seconds interval in between. 
Could you please also provide the steps to reproduce the issue. Thanks!

**Comment by indi81** [comment]
In our case we cannot reproduce this. We suspect it to be caused by Wordpress plugins writing a lot to the same files in some cache folders. We have a similar installation without certain plugins that do not get these hangs at all. We will be going to 2.7.x in march, still on 2.5.

**Comment by teizz** [comment]
this might be related to ganesha not closing fds when doing nfs4 file locks ( https://github.com/nfs-ganesha/nfs-ganesha/issues/404 )

**Comment by soumyakoduri** [comment]
@indi81 ..alright. please collect the data I requested in my earlier comment incase you run into service hang issues. Would like to check if there are any pending issues with upcall_cbk framework changes done in gfapi. Thanks!

**Comment by pasikarkkainen** [comment]
So I'm trying to understand the current status here. Does nfs-ganesha v2.7.2 help, eg. can one enable "features.cache-invalidation" when using nfs-ganesha v2.7.2 ? Thanks.

**Comment by indi81** [comment]
We went to 2.7.2 on CentOS last week and no hangs yet. Seems much more stable. 

We did immediately notice hangs with features.cache-invalidation turned on though so we disabled it again. 

Also, we were not able to collect anything as gstack locked the ganesha process totally. Have not been able to dump anything as it's a production system but will try if it shows up again.

**Comment by soumyakoduri** [comment]
> We went to 2.7.2 on CentOS last week and no hangs yet. Seems much more stable.
> 
> We did immediately notice hangs with features.cache-invalidation turned on though so we disabled it again.
> 
> Also, we were not able to collect anything as gstack locked the ganesha process totally. Have not been able to dump anything as it's a production system but will try if it shows up again.

I am not able to reproduce this issue. Could you please share your workload steps. It seems very critical to fix this issue as cache-invalidation is needed for HA.

**Comment by fbezdeka** [comment]
> We did immediately notice hangs with features.cache-invalidation turned on though so we disabled it again.

(Still the) Same problem here. (Running 2.7.1 on CentOS)

Enabling _features.cache-invalidation_ immediately leads to ganesha hangs.

The workload is still the same. 3 webservers serving some static and dynamic webapps.  
As mentioned above no traffic is needed to reproduce that.
Just turn _features.cache-invalidation_ on.

**Comment by soumyakoduri** [comment]
could you please paste the configuration (gluster volume info) of the volumes exported via nfs-ganesha?

**Comment by soumyakoduri** [comment]
and also what is the version of gluster and ganesha packages which you are using?

**Comment by fbezdeka** [comment]
gluster volume info:
``` 
Volume Name: mirror
Type: Replicate
Volume ID: 798e5d1f-...
Status: Started
Snapshot Count: 0
Number of Bricks: 1 x 3 = 3
Transport-type: tcp
Bricks:
Brick1: srv1:/glusterfs/mirror
Brick2: srv2:/glusterfs/mirror
Brick3: srv3:/glusterfs/mirror
Options Reconfigured:
transport.address-family: inet
nfs.disable: on
performance.client-io-threads: on
features.cache-invalidation: off
features.cache-invalidation-timeout: 600
performance.stat-prefetch: on
performance.cache-invalidation: on
performance.md-cache-timeout: 600
network.inode-lru-limit: 200000
nfs.rpc-auth-allow: 127.0.0.1
```

Packages:
glusterfs: 4.1.7
ganesha: 2.7.1

**Comment by soumyakoduri** [comment]
Thanks! I tested out same packages and the options..Still not able to reproduce the issue.

**Comment by pasikarkkainen** [comment]
@fbezdeka : Did you have a chance to try nfs-ganesha v2.7.2 yet?

**Comment by davplsm** [comment]
I think that I have the same behavior with versions glusterfs-5.5-1.el7.x86_64 and nfs-ganesha-2.7.2-1.el7.x86_64. The way to reproduce the problem is to launch load on the two nodes simultaneously with features.cache-invalidation enabled, for example with nfsometer.
Before freeze the following events registered in ganesha-gfapi.log:

MSGID: 104055] [glfs-fops.c:4956:glfs_cbk_upcall_data] 0-gfapi: Synctak for Upcall event_type(1) and gfid(<92>y<U+0084>~<F2>K+<82>3<9A><A6><9C>)<B1><E1>^A) failed

and finally:
C [rpc-clnt-ping.c:162:rpc_clnt_ping_timer_expired] 0-loganas-client-0: server 192.168.0.35:49153 has not responded in the last 10 seconds, disconnecting.

Probably it is also related with this entry https://www.spinics.net/lists/gluster-users/msg36042.html

Thanks

**Comment by soumyakoduri** [comment]
Thanks for the steps. I could reproduce the issue now. Will post the fix soon and update the thread.

**Comment by soumyakoduri** [comment]
The issue is in libgfapi with upcall notification processing. Have raised https://bugzilla.redhat.com/show_bug.cgi?id=1693575 and posted fix https://review.gluster.org/#/c/glusterfs/+/22436 upstream for review.

**Comment by pasikarkkainen** [comment]
@soumyakoduri Thanks a lot! So it sounds like no fix is needed in nfs-ganesha, as long as glusterfs libgfapi has your patch included?

**Comment by soumyakoduri** [comment]
Thats right!

**Comment by pasikarkkainen** [comment]
@soumyakoduri Thanks! Is there some specific procedure for requesting backports to gluster release branches? This bug seems to happen already in gluster 4.1 (and all the later versions), so the bugfix patch should be applied to all the maintained gluster branches.. (4.1.x, 5.x, 6.x).

**Comment by soumyakoduri** [comment]
yes...I am planning to backport this fix to release-6,5 and 4.1 .. will let you know once done.

**Comment by pasikarkkainen** [comment]
@soumyakoduri Awesome, thanks a lot! I can see the patch already made it to 6.x branch. Hoping to see the patch 5.x and 4.1.x branches aswell :)

**Comment by pasikarkkainen** [comment]
I can see the "gfapi: Unblock epoll thread for upcall processing" patch is now merged in all the maintained git branches: 6.x, 5.x, and 4.1.x.

glusterfs 6.1 and 5.6 versions have been already released with the patch included, and as the patch is in the 4.1.x branch, it should be included in the next 4.1.9 release aswell.

**Comment by soumyakoduri** [comment]
Thats correct. Please test those releases and let me know if the issue is resolved in your workload/setup. Thanks!

**Comment by fbezdeka** [comment]
Still waiting for the 4.1.9 release.
The changelog [1] of 4.1.8 mentioned a tentative release date: Week of 20th May, 2019

Does anyone know when we can expect it?

[1] https://docs.gluster.org/en/latest/release-notes/4.1.8/

**Comment by pasikarkkainen** [comment]
afaik there has been CI testing issues in gluster project recently, which has delayed the releases. So hopefully soon..

**Comment by soumyakoduri** [comment]
> Still waiting for the 4.1.9 release.
> The changelog [1] of 4.1.8 mentioned a tentative release date: Week of 20th May, 2019
> 
> Does anyone know when we can expect it?
> 
> [1] https://docs.gluster.org/en/latest/release-notes/4.1.8/

Hi,
Just couple of days back we uncovered another issue (my bad). Patch posted upstream for review - https://review.gluster.org/#/c/glusterfs/+/22839/

I request to hold till this patch gets merged too in the stable releases.

**Comment by pasikarkkainen** [comment]
@soumyakoduri Thanks for the heads up! I wonder if there's a chance of getting that patch to 4.1 branch? afaik the upcoming gluster 4.1.9 is the last planned release on that branch..

**Comment by soumyakoduri** [comment]
oh okay...I will check with maintainers and update

**Comment by pasikarkkainen** [comment]
Hmm, it seems v4.1.9 has been tagged already.. https://github.com/gluster/glusterfs/releases/tag/v4.1.9

**Comment by soumyakoduri** [comment]
thats true..is there any chance you can upgrade to release-5 or 6? Will backport this patch to those stable releases once merged in master.

**Comment by pasikarkkainen** [comment]
At least I should be able to update to 6.x to test. Looking forward to https://review.gluster.org/#/c/glusterfs/+/22839/ and https://review.gluster.org/#/c/glusterfs/+/22840/ getting backported for the next 6.x release. Thanks a lot!

**Comment by pasikarkkainen** [comment]
Hmm, it looks like 22839 has been merged to master and release-6 branch, but 22840 has only been merged to master, as it has a build failure in release-6 branch: https://review.gluster.org/#/c/glusterfs/+/22873/

**Comment by soumyakoduri** [comment]
There are some spurious failures in release-6 branch.. will retrigger tests

**Comment by pasikarkkainen** [comment]
@soumyakoduri : Thanks a lot! 22873 looks green now / ready to submit. Hoping it gets merged for upcoming Glusterfs 6.4 release aswell :)

**Comment by pasikarkkainen** [comment]
.. and seems to be merged to release-6 branch now. Looking forward to testing this with the upcoming glusterfs 6.4 release!

**Comment by pasikarkkainen** [comment]
It seems there will be one more 4.1.x release coming up (4.1.10) after all. Should the recent gfapi upcall fixes be backported there aswell? I mean "gfapi: fix incorrect initialization of upcall syncop arguments" and "upcall: Avoid sending notifications for invalid inodes".

**Comment by soumyakoduri** [comment]
> It seems there will be one more 4.1.x release coming up (4.1.10) after all. Should the recent gfapi upcall fixes be backported there aswell? I mean "gfapi: fix incorrect initialization of upcall syncop arguments" and "upcall: Avoid sending notifications for invalid inodes".

yes..will backport them to release-4 branch too

**Comment by pasikarkkainen** [comment]
It seems glusterfs v6.4, v5.8 and v4.1.10 have been tagged last week ( https://github.com/gluster/glusterfs/releases ), and all of them should now contain the gfapi/upcall fixes relevant to this nfs-ganesha "IO hang" issue. Please test these new versions and please write here how it works for you! Thanks.

**Comment by pasikarkkainen** [comment]
v4.1.10 release notes: https://docs.gluster.org/en/latest/release-notes/4.1.10/
v6.4 release notes: https://docs.gluster.org/en/latest/release-notes/6.4/

**Comment by soumyakoduri** [comment]
Unfortunately there is another potential issue which we discovered recently under heavy stress load - bug1733166 . There is a fix under review https://review.gluster.org/#/c/glusterfs/+/23108/. 

Will keep this thread updated regarding the backports.

**Comment by fbezdeka** [comment]
Still waiting for the 4.1.10 release to be pushed to the CentOS repos.
The build in cbs (https://cbs.centos.org/koji/buildinfo?buildID=26337) does not have a release tag yet.

@soumyakoduri Thanks for the updates.

**Comment by fbezdeka** [comment]
Sorry @soumyakoduri, but I have bad news.

I updated to 4.1.10 today. 
After enabling _features.cache-invalidation_ two of three nodes failed.

The NFS client reported that there is no response from the NFS server.
Ganesha was still running, but no longer handling any requests.

The only relevant logs i found so far:
```text
...
03/08/2019 14:48:10 : epoch 5d4570cc : srv1 : ganesha.nfsd-1709[svc_175] posix2fsal_error :FSAL :CRIT :Mapping 107(default) to ERR_FSAL_SERVERFAULT
03/08/2019 14:48:10 : epoch 5d4570cc : srv1 : ganesha.nfsd-1709[svc_138] posix2fsal_error :FSAL :CRIT :Mapping 107(default) to ERR_FSAL_SERVERFAULT
03/08/2019 14:48:10 : epoch 5d4570cc : srv1 : ganesha.nfsd-1709[svc_229] posix2fsal_error :FSAL :CRIT :Mapping 107(default) to ERR_FSAL_SERVERFAULT
03/08/2019 15:00:16 : epoch 5d4570cc : srv1 : ganesha.nfsd-1709[dbus_heartbeat] nfs_health :DBUS :WARN :Health status is unhealthy. enq new: 121183, old: 121181; deq new: 121181, old: 121181
03/08/2019 15:03:40 : epoch 5d4570cc : srv1 : ganesha.nfsd-1709[dbus_heartbeat] nfs_health :DBUS :WARN :Health status is unhealthy. enq new: 128796, old: 128795; deq new: 128795, old: 128795
03/08/2019 15:49:21 : epoch 5d4570cc : srv1 : ganesha.nfsd-1709[dbus_heartbeat] nfs_health :DBUS :WARN :Health status is unhealthy. enq new: 216306, old: 216304; deq new: 216304, old: 216304
...
```

**Comment by soumyakoduri** [comment]
> Sorry @soumyakoduri, but I have bad news.
> 
> I updated to 4.1.10 today.
> After enabling _features.cache-invalidation_ two of three nodes failed.
> 
> The NFS client reported that there is no response from the NFS server.
> Ganesha was still running, but no longer handling any requests.
> 
> The only relevant logs i found so far:
> 
> ```
> ...
> 03/08/2019 14:48:10 : epoch 5d4570cc : srv1 : ganesha.nfsd-1709[svc_175] posix2fsal_error :FSAL :CRIT :Mapping 107(default) to ERR_FSAL_SERVERFAULT
> 03/08/2019 14:48:10 : epoch 5d4570cc : srv1 : ganesha.nfsd-1709[svc_138] posix2fsal_error :FSAL :CRIT :Mapping 107(default) to ERR_FSAL_SERVERFAULT
> 03/08/2019 14:48:10 : epoch 5d4570cc : srv1 : ganesha.nfsd-1709[svc_229] posix2fsal_error :FSAL :CRIT :Mapping 107(default) to ERR_FSAL_SERVERFAULT
> 03/08/2019 15:00:16 : epoch 5d4570cc : srv1 : ganesha.nfsd-1709[dbus_heartbeat] nfs_health :DBUS :WARN :Health status is unhealthy. enq new: 121183, old: 121181; deq new: 121181, old: 121181
> 03/08/2019 15:03:40 : epoch 5d4570cc : srv1 : ganesha.nfsd-1709[dbus_heartbeat] nfs_health :DBUS :WARN :Health status is unhealthy. enq new: 128796, old: 128795; deq new: 128795, old: 128795
> 03/08/2019 15:49:21 : epoch 5d4570cc : srv1 : ganesha.nfsd-1709[dbus_heartbeat] nfs_health :DBUS :WARN :Health status is unhealthy. enq new: 216306, old: 216304; deq new: 216304, old: 216304
> ...
> ```

Thanks for the update. As I had mentioned in my last comment, this is because of bug1733166.. fix is merged in the master..waiting for backports to get merged in the stable releases -

https://review.gluster.org/#/q/I2079e95339e5d761d5060707f4555cfacab95c83

**Comment by pasikarkkainen** [comment]
glusterfs 6.5 (https://docs.gluster.org/en/latest/release-notes/6.5/) and 5.9 (https://docs.gluster.org/en/latest/release-notes/5.9/) seem to be tagged and available, with a fix for the "potential deadlock while processing callbacks in gfapi" -issue included.

**Comment by fbezdeka** [comment]
What about a backport to the 4.1. branch? 
As far as I know 4.1.10 was considered to be the last release in this branch. 
One more release to fix that issue in the 4.1. branch would be so helpful...

**Comment by soumyakoduri** [comment]
> What about a backport to the 4.1. branch?
> As far as I know 4.1.10 was considered to be the last release in this branch.
> One more release to fix that issue in the 4.1. branch would be so helpful...

I will check with release maintainer. But since 4.x is EOL'ed with upcoming gluster-7, could you consider upgrading to supported stable release?

**Comment by wyllys66** [comment]
I dont think this is just limited to gluster.  I've seen it happen when sharing a cephfs with the VFS FSAL running ganesha v2.8.0.2 on Ubuntu Linux with a 4.18.20 kernel.  Each time it occurs, we have to restart the ganesha service to clear it up.

```

19/08/2019 11:56:49 : epoch 5d2cce80 : gw01 : ganesha.nfsd-629877[dbus_heartbeat] nfs_health :DBUS :WARN :Health status is unhealthy. enq new: 220595, old: 220594; deq new: 220594, old: 220594
20/08/2019 14:36:24 : epoch 5d2cce80 : gw01 : ganesha.nfsd-629877[dbus_heartbeat] nfs_health :DBUS :WARN :Health status is unhealthy. enq new: 227785, old: 227784; deq new: 227784, old: 227784
21/08/2019 10:52:39 : epoch 5d2cce80 : gw01 : ganesha.nfsd-629877[dbus_heartbeat] nfs_health :DBUS :WARN :Health status is unhealthy. enq new: 233249, old: 233248; deq new: 233248, old: 233248
22/08/2019 07:57:20 : epoch 5d2cce80 : gw01 : ganesha.nfsd-629877[dbus_heartbeat] nfs_health :DBUS :WARN :Health status is unhealthy. enq new: 238931, old: 238930; deq new: 238930, old: 238930

```

**Comment by fbezdeka** [comment]
Our testsystem was updated to gluster 6 last week. 
The system is still in a good shape so hopefully the problem is gone.
The production system is still running on gluster 4.1 which will not receive the fix for that bug anymore.

**Comment by emanuelmanea** [comment]
Hello, @fbezdeka could you get a packet trace of the traffic that leads to the issue?
What I mean have a continous packet trace running saving into a 20G file for instance and stop that as soon as the issue is happening. Hopefully that will provide the main characteristic of the workload that brings the system in the faulty state.

**Comment by pasikarkkainen** [comment]
@emanuelmanea wasn't this bug already fixed? (fix(es) is/are in latest glusterfs 6.x release).

**Comment by fbezdeka** [comment]
@emanuelmanea, we already updated to the latest 6.x release. As @pasikarkkainen mentioned 6.x already contains the necessary fixes. 

My understanding is that the 4.1 branch is more or less EOL and no longer maintained.

I'm closing this bug now.
